tensorboard/defs/mocha_driver.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Basic TensorBoard functional tests using WebDriver."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import re
+import StringIO
+import subprocess
+import unittest
+from selenium import webdriver
+
+from selenium.webdriver.common import desired_capabilities
+from testing.web import webtest
+
+import time
+import sys
+
+ONE_SECOND = 1
+SECONDS_TO_TIMEOUT = 10
+
+class MochaTest(unittest.TestCase):
+  """Driver for the mocha_test declared in //tensorboard/defs/mocha_test.bzl.
+
+  It expects that the runfiles for this test include a _mocha_devserver
+  which is an executable. When run this executable prints console output
+  including "Listening on: [host]:[port]/", and that executable serves
+  "/index.html" which is a mocha test suite.
+
+  It runs that test suite using webdriver, and inspects the web console output
+  to see if the tests passed.
+  """
+
+  def setUp(self):
+    src_dir = os.environ["TEST_SRCDIR"]
+
+    for root, _, files in os.walk(src_dir):
+      for file in files:
+        if file.endswith("_mocha_devserver"):
+          # We found the mocha test devserver binary to run
+          binary = os.path.join(root, file)
+          break
+
+    self.process = subprocess.Popen([binary])
+    d = desired_capabilities.DesiredCapabilities.CHROME
+    d['loggingPrefs'] = { 'browser':'ALL' }
+
+    self.driver = webtest.new_webdriver_session(capabilities=d)
+    self.driver.get("http://localhost:6006/index.html")
```

Comment:
Basically if it can't bind to 6006 then what it does right now is it keeps incrementing the port number until it finds something that works. Then it prints the URL with the port number to the console. What I recommend doing is reading the output of the Popen, getting the port number from there, and that way you'll also know the server is ready.